feat: Adding prototype of Device Loan Ledger lab

kyaaron opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [X ] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [X ] My pull request has a [descriptive title](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request?id=prepare-a-good-pr-title) (**not** a vague title like `Update index.md`)

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

References issue https://github.com/freeCodeCamp/freeCodeCamp/issues/64118

<!-- Feel free to add any additional description of changes below this line -->
This is a prototype for a new lab for the Full Stack Certification Curriculum, as part of Naomi's sprints group for November-December. This lab has the camper working with JavaScript objects to update a device ledger for Checked In and Checked Out devices.

This lab will be part of the JavaScript Objects section of the course. It will be right after the Recipe Tracker workshop and before the Quiz Game lab.

<img width="611" height="402" alt="image" src="https://github.com/user-attachments/assets/95d0010d-60a4-4bd2-82fb-f9f0942c8941" />

jdwilkin4 reviewed (commented):
Can you update your post to say where this is going to go in the curriculum order? That will help use review it so we know what topics this lab should cover

jdwilkin4 commented on fullstack-cert/js-projects/lab-device-loan-ledger/user_stories.md:
```diff
@@ -0,0 +1,6 @@
+# User Stories
+1. You should model the ledger as `{ [assetTag]: { type, status, borrower: { name, email }, dueDate } }`.
+2. You should implement `checkoutDevice(ledger, assetTag, borrower)` that clones state, sets borrower info, and logs confirmations.
```
I spot two things here.

The phrase "clones state" might not work for the level of camper at this stage of the curriculum. So I would suggest rewording that to match their current skill level. 

maybe something like `makes a copy of the current data`? or something along those lines. 

Also, the logs confirmations part, it would be easier from a testing perspective if the `checkoutDevice` returned a string. Then you can call the `checkoutDevice` and log that result to the console. that will be an easier regex test then having the console.log inside of the function like it is currently.

kyaaron replied:
Updated the wording from "clones state" to "copies the ledger object" instead, and I updated my `checkoutDevice()` function to return a string instead of logging to the console.

jdwilkin4 commented on fullstack-cert/js-projects/lab-device-loan-ledger/user_stories.md:
```diff
@@ -0,0 +1,6 @@
+# User Stories
+1. You should model the ledger as `{ [assetTag]: { type, status, borrower: { name, email }, dueDate } }`.
```
I would suggest adding a couple of sentences before this explaining what type of project they will build. Since this lab has no HTML or CSS and hence no preview showing what is happening, campers will need more context to build out this lab.

kyaaron replied:
Done! I created a scenario for the camper under a new section in the .md file called "Lab Description" which includes the structure of the equipmentLedger object and a note explaining that `assetTag` is an integer and the rest of the child object property values are strings.

jdwilkin4 commented on fullstack-cert/js-projects/lab-device-loan-ledger/user_stories.md:
```diff
@@ -0,0 +1,6 @@
+1. You should model the ledger as `{ [assetTag]: { type, status, borrower: { name, email }, dueDate } }`.
+2. You should implement `checkoutDevice(ledger, assetTag, borrower)` that clones state, sets borrower info, and logs confirmations.
+3. You should implement `checkinDevice(ledger, assetTag)` that clears borrower info and updates status.
+4. You should implement `listOverdueDevices(ledger, today)` that returns an array of overdue entries sorted by due date.
```
Not sure about this user story. Even though your solution doesn't rely on the Date object, I can still see campers reading this and thinking they have to use it. And then being confused because it hasn't been taught yet. 

Not sure if we should remove it fore that reason or find another way to reword it to make it clear that the date object shouldn't be used.

kyaaron replied:
I was wondering this as well but went with it. I'll ping Naomi in Discord to see.

11969545395k-art replied:
404

jdwilkin4 commented on fullstack-cert/js-projects/lab-device-loan-ledger/script.js:
```diff
@@ -0,0 +1,125 @@
+function loadLedger(json) {
+    json.parse();
+    return 0;
+}
```
are these being used somewhere? They are mentioned in the user stories, but I don't see where we are actually calling them.

kyaaron replied:
Added these helper functions to the main functions in the lab. I also updated the starting ledger object to a string in the .js file to mimic a JSON file, so the helper functions would work.

jdwilkin4 reviewed (commented):
I left some comments regarding some of the user stories. But also, I noticed that most of these functions and `equipmentLedger` object isn't being used anywhere. 

Even though campers won't see this final solution and will implement their own solutions, we still want to make sure all variables and functions are being utilized some where. Whether that be through logging that function call or doing something else with it.

kyaaron commented on fullstack-cert/js-projects/lab-device-loan-ledger/user_stories.md:
```diff
@@ -0,0 +1,6 @@
+2. You should implement `checkoutDevice(ledger, assetTag, borrower)` that clones state, sets borrower info, and logs confirmations.
+3. You should implement `checkinDevice(ledger, assetTag)` that clears borrower info and updates status.
+4. You should implement `listOverdueDevices(ledger, today)` that returns an array of overdue entries sorted by due date.
+5. You should implement `serializeLedger(ledger)` and `loadLedger(json)` helpers for persistence.
```
Done. Updated US-5 to:

"You should write `serializeLedger(ledger)` and `loadLedger(json)` helpers functions to load (convert) a JSON file into a JavaScript object, or save (serialize) a JavaScript object into a JSON object. _Hint: Refer back to your JSON lessons preceeding this lab if you need help._"

kyaaron commented:
@jdwilkin4 I made the updates from your most recent review. Below is a list of the changes made:
- Removed Quincy's name in the example object and replaced with another test name
- I updated the user stories in the README.md file to better explain load and serialize and what the functions should accomplish.
- I updated the README.md file with a scenario to better explain the lab and tasks the camper will need to do. I also specified the data type of the `assetTag` once it is parsed.
- I corrected my helper functions and inserted them into the checkout, checkin, and listOverDue functions. I tested the updates to make sure they worked. Related to this, I also updated the starter equipmentLedger from an object to a string data type.

I left specific comments in each comment you gave about specific values. Please let me know if you have any questions or if I should update anything else. Thank you!

kyaaron commented:
Also, updated the console.log() statements to return statements to make it easier to test.

jdwilkin4 reviewed (approved):
Going to go ahead and approve this.
There will probably need to be tweaks to the user stories based on how testing goes in the main repo. If any changes need to be made, it will happen in the next review stage in the main repo.
